test/pylib/manager_client.py

```diff
@@ -47,15 +47,26 @@ def __init__(self, sock_path: str, port: int, use_ssl: bool, auth_provider: Any|
         self.ccluster: Optional[CassandraCluster] = None
         self.cql: Optional[CassandraSession] = None
         # A client for communicating with ScyllaClusterManager (server)
-        self.client = UnixRESTClient(sock_path)
+        self.sock_path = sock_path
+        self.clients_pool = {asyncio.get_running_loop(): UnixRESTClient(sock_path)}
         self.api = ScyllaRESTAPIClient()
         self.metrics = ScyllaMetricsClient()
         self.thread_pool = ThreadPoolExecutor()
 
+    @property
+    def client(self):
+        _client = self.clients_pool.get(asyncio.get_running_loop(), None)
+        if _client is None:
+            _client = UnixRESTClient(self.sock_path)
+            self.clients_pool[asyncio.get_running_loop()] = _client
+        return _client
+
     async def stop(self):
         """Close driver"""
         self.driver_close()
-        await self.client.shutdown()
+        # TODO: good candidate for safe_gather  https://github.com/scylladb/scylladb/pull/17781
+        #  to make sure tha all connections is closed
+        await asyncio.gather(*[client.shutdown() for client in self.clients_pool.values()])
```

Comment:
I don't know much about asyncio so I have to ask: Why is it fine that we use asyncio.gather in one event loop (the one running this code) to wait for events in other event loops?
Is the lifetime even guaranteed (e.g., is it possible that some other event loop has already exited)?

Reply:
> I don't know much about asyncio so I have to ask: Why is it fine that we use asyncio.gather in one event loop (the one running this code) to wait for events in other event loops? Is the lifetime even guaranteed (e.g., is it possible that some other event loop has already exited)?

here I'm not  wait for events in other event loops. I have a list of client objects,  spawn new COROs (client.shutdown()) and waiting for it. I'm not working with other event loops at all

and it is possible to create a use object in different even loops( it is just bad practice to create multiple event_loops in one thread)

the problem is aiohttp in Python 3.12 has a check that forbids establishing (creating) ```UnixRESTClient``` and calling ```request``` in different event_loops. I guess this exists due to ```request``` implementation logic

this check is here  /aiohttp/helpers.py. class TimerContext
```
    def __enter__(self) -> BaseTimerContext:
        task = current_task(loop=self._loop)

        if task is None:
            raise RuntimeError(
                f"Timeout context manager should be used inside a task {id(self._loop)} {id(asyncio.get_running_loop())}"
            )

        if self._cancelled:
            raise asyncio.TimeoutError from None

        self._tasks.append(task)
        return self
```
but call ```shutdown```  if different event_loops is allowed by current aiohttp logic.

Reply:
Ok, so it sounds like I mis-described the problem but did get it approximately right: Creating a client on one loop and calling request() in another loop is not allowed, but calling shutdown() in the other loop is allowed. Isn't this completely arbitrary?
Maybe we should just accept your patch as fixing some aiohttp quirk that has no logical explanation, but we need to do it anyway to appease the library :-(

Reply:
the problem comes from mismatching  aiohttp usage rules and several fixtures in pytest


event_loop fixture in /test/pylib/async_cql.py that creates new event loop

```
@pytest.fixture(scope="session")
def event_loop(request):
    """Change default pytest-asyncio event_loop fixture scope to session to
       allow async fixtures with scope larger than function. (e.g. manager fixture)
       See https://github.com/pytest-dev/pytest-asyncio/issues/68"""
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    loop.close()
```

and 2 fixtures in cinftest.py manager_internal(scope session) and manager(scope function)

if remove event_loop fixture and set manager_internal scope function it will be 1 event_loop and 1 ManagerClient per test and everything will work fine. 

but now manager_internal and test body runs in the event pool A, manager(and random_tables that also use manager) event pool B.(next test e event pool  will be A(same) and C(new) )

Reply:
Ok. I'm not sure that I have understood all the details, but I'm now convinced that you have :-)
Now the main thing left is to understand why @tchaikov said above that your patch does **not** fix the problem he had.